src/main/kotlin/com/cjbooms/fabrikt/generators/controller/KtorControllerInterfaceGenerator.kt

```diff
@@ -0,0 +1,362 @@
+            val typeName = param.type.copy(nullable = false) // not nullable because we handle that in the queryParameters.get* below
+            if (param.isRequired) {
+                codeBlock.appendLine("val ${param.name} = call.request.queryParameters.getOrFail<$typeName>(\"${param.originalName}\")")
+            } else {
+                codeBlock.appendLine("val ${param.name} = call.request.queryParameters.getTyped<$typeName>(\"${param.originalName}\")")
+            }
+        }
+
+        bodyParams.forEach { param ->
+            codeBlock.appendLine("val ${param.name} = call.receive<${param.type.simpleName()}>()")
+        }
+
+        val methodParameters =
+            listOf(headerParams, pathParams, queryParams, bodyParams).flatten()
+                .joinToString(", ") { it.name }
+                .let {
+                    if (it.isNotEmpty()) "call, $it"
+                    else "call"
+                }
+
+        codeBlock.appendLine("val result = controller.$methodName($methodParameters)")
+
+        if (happyPathResponse.simpleName() == Unit::class.simpleName) {
+            // When return type is Unit we only respond with the status code.
+            // Note however that in some cases the controller may choose to respond directly on the call
+            // in which case that takes precedence in Ktor's routing. For example: call.respondRedirect().
+            codeBlock.appendLine("call.respond(result.status)")
+        } else {
+            codeBlock.appendLine("call.respond(result.status, result.message)")
+        }
+
+        codeBlock.appendLine("}")
+
+        if (addAuth) {
+            codeBlock.appendLine("}")
+        }
+
+        return codeBlock.toString()
+    }
+
+    override fun generateLibrary(): Collection<ControllerLibraryType> = listOf(
+        buildRoutingUtils(),
+        buildControllerResult()
+    )
+
+    /**
+     * Builds a data class to be used as a return type for controller methods
+     */
+    private fun buildControllerResult(): ControllerLibraryType {
+        val genericT = TypeVariableName("T")
+
+        val spec = TypeSpec.classBuilder(CONTROLLER_RESULT_CLASS_NAME)
+            .addModifiers(KModifier.DATA)
+            .addTypeVariable(genericT)
+            .primaryConstructor(
+                FunSpec.constructorBuilder()
+                    .addParameter("status", ClassName("io.ktor.http", "HttpStatusCode"))
+                    .addParameter("message", genericT) // naming aligned with Ktor's call.response(status, message)
+                    .build()
+            )
+            .addProperty(
+                PropertySpec.builder("status", ClassName("io.ktor.http", "HttpStatusCode"))
+                    .initializer("status")
+                    .build()
+            )
+            .addProperty(
+                PropertySpec.builder("message", genericT)
+                    .initializer("message")
+                    .build()
+            )
+            .build()
```

Comment:
I think this is too limited of a response struct, considering what Ktor offers in general: https://ktor.io/docs/server-responses.html

Reply:
Yeah I agree it is too limited to cover all use cases, and this is why I have chosen to pass `ApplicationCall` to the controllers which allows them to both read from the request, but also respond in the various ways Ktor [supports](https://ktor.io/docs/server-responses.html).

Unfortunately this leads to an ambiguous API where the controller method is expected to return a ControllerResult<Unit>, but is also free to "close" the call with e.g. `call.respondRedirect`. 

In practise this leads to the following implementation:
```kotlin
call.respondRedirect("https://some.other.domain/path")
return ControllerResult(HttpStatusCode.OK, Unit)
```
Ktor will respond with the first `respond` and the `ControllerResult` will be ignored.

In the initial version I had simply skipped the ControllerResult in cases where the spec does not mention a return model. This would avoid the awkward code mentioned above.

Possible alternatives would be to
1. Not force a return type and let the controller call `call.respondX`. Here we loose the type safety of the return type, which is not desirable. 
2. Provide a return struct that supports all the same `respond` as Ktor. This would have to be updated whenever Ktor changes.
3. Other alternatives?

Reply:
That is a dilemma :)

What about creating two functions for each endpoint; a "simple" and a "powerful" one.

```kotlin

interface MyController {
    // Only returns 200 with the returned body; or 500 on an exception.
    public suspend fun post(
        querySomeObject: String,
        bodySomeObject: SomeObject,
    ): Unit
    
    // Can use arbitrary returns.
    public suspend fun post(
        call: ApplicationCall,
        querySomeObject: String,
        bodySomeObject: SomeObject,
    ) {
        val responseBody = post(querySomeObject, bodySomeObject)
        call.respond(responseBody)
        // ...
    }
    
    // Routing ...
}
```

The only issue I see is if users override both functions for one endpoint, but I feel like that would be a rare mistake and not that difficult to catch.

Of course, it is arguable whether the first method actually serves any purpose or if 99% of use cases require the second one anyway.

Reply:
Having looked more into this, I would like to go with simply not requiring the implementation to return `ControllerResult` in cases where the route has no return content. 

In that case Ktor will respond with `200 OK` by default, but we still allow the implementation to respond with a e.g. a redirect if needed.

This change has been added in https://github.com/cjbooms/fabrikt/pull/280/commits/35c3195e9a43e1a3498a59ca5201d5078988e8a3.

Reply:
Turns out I was wrong about the default: if no response is provided Ktor will respond with `404 Not Found` 😬 

@atollk Would you be OK with making routes without response content return `HttpStatusCode`? If the implementation choses to perform an early `call.respond` the code is a little ambiguous, but that is already the case with the endpoints that return content. Here the implementation would still be required to return a `ControllerResult` even though it performs an early respond.

```kotlin
interface Controller {
  suspend fun testPath(call: ApplicationCall): HttpStatusCode
  
  companion object {
    get("/test") {
      val resultStatus = controller.testPath(call)
      call.respond(resultStatus)
    }
  }
}

// respond with 200 OK

override suspend fun testPath(call: ApplicationCall): HttpStatusCode {
  return HttpStatusCode.OK
}

// or redirect

override suspend fun testPath(call: ApplicationCall): HttpStatusCode {
  call.respondRedirect("/someOtherPath")
  return HttpStatusCode.InternalServerError // <-- should never be reached
}
```

Reply:
I don't know, I am not a fan of forcing the implementation to have a return even if will never be reached; that'd be akin to forcing users to have a `return Unit` in all their functions which don't actually return anything.

I think that a function without any return that forces the user to implement any responses using the `ApplicationCall` object directly is the best way. It's the only option that provides generic support for all use cases and that's the option that should be included in a tool like Fabrikt imo. Special convenience options could be added in later revisions in theory.

Reply:
I agree that it is not nice to force a return that will never be reached.

I like the design where we provide a simple and an advanced option, and I think it could cover all use cases. Any time the user wishes to deal directly with `call` (either to read additional information from the request, or to provide a custom response) he would implement the advanced function.

I am thinking it could look like this, where the default implementation of the advanced function can be overriden by the user:

```kotlin
public interface InternalEventsController {

  // simple version
  public suspend fun post(bulkEntityDetails: BulkEntityDetails): ControllerResult<EventResults>

  // advanced version; user is responsible for returning a response
  public suspend fun post(bulkEntityDetails: BulkEntityDetails, call: ApplicationCall) {
    // default implementation delegates to the simple version
    val result = post(bulkEntityDetails)
    call.respond(result.status, result.message)
  }

  public companion object {
    public fun Route.internalEventsRoutes(controller: InternalEventsController) {
      post("/internal/events") {
        val bulkEntityDetails = call.receive<BulkEntityDetails>()
        controller.post(bulkEntityDetails, call)
      }
    }
}
```

In cases where no model is returned by the endpoint we could simply return HttpStatusCode from the simple function:

```kotlin
public interface InternalEventsController {

  public suspend fun post(bulkEntityDetails: BulkEntityDetails): HttpStatusCode

  public suspend fun post(bulkEntityDetails: BulkEntityDetails, call: ApplicationCall) {
    val status = post(bulkEntityDetails)
    call.respond(status)
  }

  public companion object {
    public fun Route.internalEventsRoutes(controller: InternalEventsController) {
      post("/internal/events") {
        val bulkEntityDetails = call.receive<BulkEntityDetails>()
        controller.post(bulkEntityDetails, call)
      }
    }
}
```

Does that resonate with you?

Reply:
I just realised that the above will not cut it: The user will be forced to always implement the simple version as it is abstract 🤔 

We could however provide a default implementation which throws in case it has not been overridden. This will happen in cases where neither the advanced nor the simple version has been implemented.

Downside is that the user will not get warned at compile time if they have forgotten to implement one of the routes.

I am all ears if you have alternative solutions i mind.

```kotlin
public suspend fun post(bulkEntityDetails: BulkEntityDetails): ControllerResult<EventResults> {
  throw BadRequestException("Not implemented")
}

public suspend fun post(bulkEntityDetails: BulkEntityDetails, call: ApplicationCall) {
  // default implementation delegates to the simple version
  val result = post(bulkEntityDetails)
  call.respond(result.status, result.message)
}
```

Reply:
> I just realised that the above will not cut it: The user will be forced to always implement the simple version as it is abstract 🤔

True, I didn't think about that either. 

The `NotImplementedError` is not great, as you said, because it's lacking compile checks but that would also have been my idea at the moment. Maybe also respond a 501 status code to the call.

I'm still not to keen on the `ControllerResult` custom class. Maybe we should create a feature request on the Ktor tracker to see if Jetbrains would be willing to add a `Response`-like type that we could use. Fabrikt isn't the first time I wish something like that existed :)

But for the time being, if you want to merge the MR, it might be the right choice to just offer the "advanced" version for now and potentially add convenience overloads in the future? We could add a comment about the excepted return type to make it slightly less annoying to the user.

Reply:
Would having an extension property help resolve this in any way? Eg, allow the user to specify the return type desired for each operation

x-ktor-response-type: blah

I'm not fully following this PR, so feel free to disregard this comment if not helpful

Reply:
Let us stick with offering a function which delegates responding to the implementation (aka the advanced version) 👍 

I have added extra documentation to the generated methods which informs the user about the responsibility to respond to the call.

```kotlin
public interface InternalEventsController {
    /**
     * Generate change events for a list of entities
     *
     * Route is expected to respond with [examples.githubApi.models.EventResults].
     * Use [io.ktor.server.response.respond] to send the response.
     *
     * @param bulkEntityDetails
     * @param call The Ktor application call
     */
    public suspend fun post(bulkEntityDetails: BulkEntityDetails, call: ApplicationCall)

    public companion object {
        /**
         * Mounts all routes for the InternalEvents resource
         *
         * - POST /internal/events Generate change events for a list of entities
         */
        public fun Route.internalEventsRoutes(controller: InternalEventsController) {
            post("/internal/events") {
                val bulkEntityDetails = call.receive<BulkEntityDetails>()
                controller.post(bulkEntityDetails, call)
            }
        }
        // ...
    }
}
```

With this change the `ControllerResult` class as well as the `generateLibrary` function on the controller generator has been made obsolete, which is why I have removed them again.

Following this decision I believe the PR is ready to be merged, @cjbooms 🚀